### PR TITLE
Tictactoe

### PR DESCRIPTION
**Pull request by minqianlu**

### CODE REVIEW

File: config.ini [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ; config.ini
 
 [Game]
-name = chess
+name = tictactoe
 ; chess, go, othello
```
Comment (zdefranc) [inline config.ini]: Change go to "tictactoe"

File: main.py [hunk 2]
```diff
@@ -39,13 +43,27 @@ def main():
     best_individual = None
     if algorithm_name == "genetic_algorithm":
         # Create a GeneticAlgorithm instance
+
+        # Evolve the population for a certain number of generations
+        if game_name == "tictactoe":
```
Comment (BenKurrek) [inline main.py]: Fix please

File: main.py [hunk 3]
```diff
@@ -27,6 +29,8 @@ def main():
 
     # Read configuration from the file
     config = read_config()
+    print(f"{Fore.MAGENTA}~~~~~~~~~~~~~~~~~~~~   STARTING ALGORITHM   ~~~~~~~~~~~~~~~~~~~~{Style.RESET_ALL}")
```
Comment (zdefranc) [inline main.py:32]: Remove

File: main.py [hunk 4]
```diff
@@ -39,13 +43,27 @@ def main():
...
+        #genetic_algorithm.plot_evolution_history()
         genetic_algorithm = GeneticAlgorithm(game_name, population_size=20, mutation_rate=0.8, seed=seed)
 
         # Evolve the population for a certain number of generations
-        best_individual = genetic_algorithm.evolve(generations=200)
+        best_individual = genetic_algorithm.evolve(generations=generations)
         genetic_algorithm.plot_evolution_history()
     elif algorithm_name == "pso":
-        pso = PSO(game_name, num_particles=10)
+        if game_name == "tictactoe":
```
Comment (BenKurrek) [inline main.py]: fix please

File: src/game/tictactoe/pawn.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
+# pawn.py
+
+import chess
+from chess import Square, Piece
+
+class PawnEvaluator:
```
Comment (BenKurrek) [inline src/game/tictactoe/pawn.py]: remove this file please

File: main.py [hunk 6]
```diff
@@ -39,13 +43,27 @@ def main():
     best_individual = None
     if algorithm_name == "genetic_algorithm":
         # Create a GeneticAlgorithm instance
+
+        # Evolve the population for a certain number of generations
+        if game_name == "tictactoe":
+            genetic_algorithm = GeneticAlgorithm(game_name, population_size=4, mutation_rate=0.5)
+            generations = 20
```
Comment (zdefranc) [inline main.py]: Generations should be set individually remove if statement

File: main.py (hunk 4 above)
Comment (zdefranc) [inline main.py]: Remove tictactoe only if statement

File: src/game/tictactoe/ttt_game.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+
+    def get_weights(self):
+        return self.weights
+    
+    def get_best_move(self):
+        best_move, best_score = self.determine_best_move_from_weights()
+        return best_move
+    
+    def rank_moves(self):
+        self.moves_and_scores = sorted(self.moves_and_scores, key=lambda x: x[1], reverse=True)
```
Comment (BenKurrek) [inline src/game/tictactoe/ttt_game.py:71]: you didn't change this to a proper data structure... this is innefficient. Normally I would say change it but tic tac toe is not very intensive so it's not that big a deal.

File: src/optimization/genetic_algorithm.py [hunk 8]
```diff
@@ -80,6 +84,12 @@ def initialize_population(self):
             board_data = extract_random_chess_positions(num_positions=1, seed=self.seed)[0]
             # Create the population given the set of initial individuals
             return [create_base_game(self.game_name, board_data) for _ in range(self.population_size)]
+        elif self.game_name == "tictactoe":
+            print("Tic Tac Toe")
+            board_data = extract_random_ttt_positions(num_positions=1)[0]
+            print(board_data)
```
Comment (BenKurrek) [inline src/optimization/genetic_algorithm.py]: remove

File: src/optimization/genetic_algorithm.py [hunk 9]
```diff
@@ -80,6 +84,12 @@ def initialize_population(self):
...
+            return[create_base_game(self.game_name, board_data) for _ in range(self.population_size)]
+            # self.game = Othello( )  # Replace with actual Othello initialization
```
Comment (BenKurrek) [inline src/optimization/genetic_algorithm.py]: remove

File: src/optimization/genetic_algorithm.py [hunk 10]
```diff
@@ -153,6 +165,12 @@ def evolve(self, generations, target_fitness=None) -> BaseGame:
             self.population = new_population
 
             best_move_rank = best_individual.rank_move(best_move)
+            if not (generation + 1) % 10:
```
Comment (BenKurrek) [inline src/optimization/genetic_algorithm.py]: remove logs

File: src/utility/chess_extraction.py [hunk 11]
```diff
@@ -1,15 +1,15 @@
...
 from tqdm import tqdm
-import time
+from chess import engine, pgn
 
 # Evaluate the score of every legal move (and get the next best move)
 def stockfish_evaluation(board):
     two_ply_moves = 4
 
-    engine = chess.engine.SimpleEngine.popen_uci("./stockfish")
+    myEngine = engine.SimpleEngine.popen_uci("./stockfish")
```
Comment (BenKurrek) [inline src/utility/chess_extraction.py]: revert please

**Review by BenKurrek — CHANGES_REQUESTED** [review]

File: src/game/tictactoe/ttt_game.py [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+
+    # Random starting genes for the chromosome based on lower and upper bounds
+    def initialize_random_weights(self):
+        self.weights = [random.uniform(float(lower), float(upper)) for lower, upper in weight_bounds]
+
+    def get_board_data(self):
+        return [self.board, self.moves_and_scores]
+
+    def rank_move(self, move):
+        rank = 1
```
Comment (zdefranc) [inline src/game/tictactoe/ttt_game.py]: Best move should be 0 according to every other game convention

File: src/optimization/genetic_algorithm.py [hunk 13]
```diff
@@ -117,13 +127,15 @@ def crossover(self, game1, game2):
 
     def evolve(self, generations, target_fitness=None) -> BaseGame:
         for generation in range(generations):
+            if not (generation + 1) % 10:
+                print(f"{Fore.CYAN}~~~~~~~~~~~~~~~~~~~~   GENERATION: {generation + 1}   ~~~~~~~~~~~~~~~~~~~~{Style.RESET_ALL}")
```
Comment (zdefranc) [inline src/optimization/genetic_algorithm.py]: Should this be here?

File: src/optimization/genetic_algorithm.py (hunk 10 above)
Comment (zdefranc) [inline src/optimization/genetic_algorithm.py]: Is this necessary?

File: src/optimization/genetic_algorithm.py [hunk 14]
```diff
@@ -161,8 +179,8 @@ def evolve(self, generations, target_fitness=None) -> BaseGame:
             })
             self.weight_history.append(best_individual.get_weights())
 
-            print(f"\nGeneration {generation + 1}, Best Fitness: {best_fitness_score}, Best Move: {best_move} with rank: {best_move_rank}")
-            print(f"Weights: {best_individual.get_weights()}\n")
+            if not (generation + 1) % 10:
```
Comment (zdefranc) [inline src/optimization/genetic_algorithm.py]: Is this necessary?

**Pull request merged**